File: NamelessInteractive.Insured.Domain/Individual.fs
module NamelessInteractive.Insured.Domain.Individual

open System
open NamelessInteractive.Insured.Domain.Shared
open System.ComponentModel.DataAnnotations

type Race = 
    {
        Id: Identifier
        Code: string
        Description: string
    }

type MaritalStatus =
    {
        Id: Identifier
        Code: string
        Description: string
    }

type Gender = 
    {
        Id: Identifier
        Code: string
        Description: string
    }

type Salutation = 
    {
        Id: Identifier
        Code: string
        Description: string
    }

type IdentityNumber = IdentityNumber of IdentityNumber: string
type PassportNumber = PassportNumber of PassportNumber: string
type IndividualIdentityType =
    | IdentityNumberOnly of IdentityNumber: IdentityNumber
    | PassportNumberOnly of PassportNumber: PassportNumber
    | IdentityNumberAndPassportNumber of IdentityNumber: IdentityNumber * PassportNumber: PassportNumber

type PersonName = 
    {
        [<Display(Name="First Name")>]
        [<Required>]
        FirstName: string
        [<Display(Name="Middle Name")>]
        MiddleName: string option
        [<Display(Name="Last Name")>]
        [<Required>]
        LastName: string
        [<Display(Name="Maiden Name")>]
        MaidenName: string option
    }

type Individual =
    {
        Id: Identifier
        [<Display(Name="Name")>]
        Name: PersonName
        [<Display(Name="Date of Birth")>]
        DateOfBirth: DateTime
        [<Display(Name="Race")>]
        Race: Race option
        [<Display(Name="Marital Status")>]
        MaritalStatus: MaritalStatus option
        [<Display(Name="Gender")>]
        Gender: Gender option
        [<Display(Name="Salutation")>]
        Salutation: Salutation option
        [<Display(Name="Identity Number")>]
        IdentityNumber: IdentityNumber option
        [<Display(Name="Passport Number")>]
        PassportNumber: PassportNumber option
    }

File: NamelessInteractive.Insured.Web/Scripts/App/Individual/ViewModels/Individual.ts
module InsuredApp {
    export module ViewModels {
        export module Individual {
            export class Individual {
                Id: Core.Identifier;
                Name: PersonName;
                DateOfBirth: Date;
                Race: Race;
                MaritalStatus: MaritalStatus;
                Gender: Gender;
                Salutation: Salutation;
                IdentityNumber: IdentityNumber;
                PassportNumber: PassportNumber;
                Lookups: IndividualLookup;
                constructor() {
                    this.Lookups = new IndividualLookup();
                }

                static IfNull(a, b, c) {
                    return Utilities.Shared.ParseUtilities.IfNull(a, b, c);
                }

                static Parse(data: Individual) {
                    var result = new Individual();
                    result.Id = new Core.Identifier(data.Id.Identifier);
                    result.Name = PersonName.Create(data.Name.FirstName, data.Name.MiddleName, data.Name.LastName, data.Name.MaidenName);
                    result.DateOfBirth = data.DateOfBirth;
                    result.Race = Race.Create(data.Race.Id.Identifier, data.Race.Code, data.Race.Description);
                    result.MaritalStatus = MaritalStatus.Create(data.MaritalStatus.Id.Identifier, data.MaritalStatus.Code, data.MaritalStatus.Description);
                    result.Gender = Gender.Create(data.Gender.Id.Identifier, data.Gender.Code, data.Gender.Description);
                    result.Salutation = Salutation.Create(data.Salutation.Id.Identifier, data.Salutation.Code, data.Salutation.Description);
                    if (data.IdentityNumber != null && data.IdentityNumber != undefined) {
                        result.IdentityNumber = new IdentityNumber(data.IdentityNumber.IdentityNumber);
                    }
                    if (data.PassportNumber != null && data.PassportNumber != undefined) {
                        result.PassportNumber = new PassportNumber(data.PassportNumber.PassportNumber);
                    }
                    return result;
                }
            }
        }
    }
} 